core/field_dropdown.js

```diff
@@ -531,6 +532,28 @@ Blockly.FieldDropdown.prototype.renderSelectedText_ = function() {
       Blockly.Field.X_PADDING;
 };
 
+/**
+ * @override
+ */
+Blockly.FieldDropdown.prototype.getText_ = function() {
+  if (this.selectedOption_) {
+    if (typeof this.selectedOption_ == 'object') {
+      return this.selectedOption_['alt'];
+    }
+    return this.selectedOption_;
+  }
+  return null;
+};
+
+/**
+ * Set the selected option on this dropdown.
+ * @param {string|Object} option The option to select.
+ * @protected
+ */
+Blockly.FieldDropdown.prototype.setSelectedOption_ = function(option) {
```

Comment:
Yep, thanks for pointing out the second note. I'm a big advocate for having better types, so since there are very specific properties we're looking for. We can create a type for that, similar to: https://github.com/google/blockly/blob/develop/core/theme.js#L38

As for your first point, this is selected the key on the options list. Field options are a name -> id (value) map.
I've updated the jsdoc to indicate what the option is, which is the ``human-readable text or image``.